gossip/infostore.go

```diff
@@ -237,11 +241,34 @@ func (is *infoStore) processCallbacks(key string, content roachpb.Value) {
 		}
 	}
 
-	// Run callbacks in a goroutine to avoid mutex reentry.
-	go func() {
+	// Add the matching callbacks to the callback work list.
+	f := func() {
 		for _, method := range matches {
 			method(key, content)
 		}
+	}
+	is.callbackMu.Lock()
+	is.callbackWork = append(is.callbackWork, f)
+	is.callbackMu.Unlock()
+
+	// Run callbacks in a goroutine to avoid mutex reentry. We also guarantee
+	// callbacks are run in order such that if a key is updated twice in
+	// succession, the second callback will never be run before the first.
+	go func() {
+		// Grab the callback semaphore to prevent another goroutine from invoking
```

Comment:
I'm using semaphore in the sense of a gate for a limited resource (the number of goroutines invoking callbacks). More idiomatic would be to use a channel of length 1, but a mutex is simpler here. Note I don't want to hold `callbackMu` while invoking the callbacks as doing so leads to potential mutex reentrancy issues that we're trying to avoid. Let me see if I can come up with a better term than semaphore.